### ISSUE TITLE
-split-bookmarks into UTF-8 "@B" titles

### ISSUE THREAD

**Issue by jlcd** [coherentgraphics/cpdf-binaries#30]
I've read that the `-split-bookmarks` operation removes some characters, as per:

```
The bookmark text used for a name is converted from unicode to 7 bit ASCII, and the following
characters are removed, in addition to any character with ASCII code less than 32:
/ ? < > \ : * | " ˆ + =
```

Not sure why it was made this way, but are UTF-8 bookmark titles expected to be implemented in the near future?
If not, may I leave this open as a feature request?

**Comment by jlcd** [comment]
Man you've a quick response.

Now I realize why you did this. Some characters may not be filename-friendly, so I agree 100% with you.
I just now found out about the `-utf8` flag (when I read about listing bookmarks). From my [limited] context on cpdf, I guess this would be indeed the flag to use when keeping bookmark-filenames [mostly] unchanged. It would be something like "force utf-8 filenames as I'm aware of the consequences".

Of course I don't expect this to be done lightning fast, but do you have an ETA for when this will reach a stable version?
Asking just to know if I should work on some workarounds or wait.

And of course, many thanks for this excelent tool.

**Comment by jlcd** [comment]
Not sure if this is the proper way to generate UTF8 filenames (when the `-utf8` flag is set), but I gave it a try: https://github.com/jlcd/cpdf-source/commit/a5e9f4dcc5e56cc1afd7fa493ed7680dad755f22#diff-a1ea83527a319a64f1e227a3add40e68

Never develop anything with OCaml, so I'm pretty sure there are some bits off place there.

Seems to be working for my scenario.

**Comment by johnwhitington** [comment]
Thanks! I'll take a detailed look soon.

(Speed: you somehow built the bytecode version not the native code version?)

**Comment by jlcd** [comment]
Not sure, I just ran `make` to compile it.

Should I have done it in any other way?

**Comment by johnwhitington** [comment]
Can you give me the output of `file cpdf`?

**Comment by jlcd** [comment]
Ok, just confirming, it really is way slower:

```
root@21b14f4a4c40:/tmp/cpdf-source# time ./cpdf -split-bookmarks 0 ../x.pdf -utf8 -o ../my/$RANDOM%%%%%@B.pdf

real    0m21.702s
user    0m17.220s
sys 0m1.890s
root@21b14f4a4c40:/tmp/cpdf-source# time ../cpdf -split-bookmarks 0 ../x.pdf -utf8 -o ../my/$RANDOM%%%%%@B.pdf

real    0m6.590s
user    0m4.600s
sys 0m0.310s
root@21b14f4a4c40:/tmp/cpdf-source# 
```

And the `file <slower_cpdf>` command output you asked:
```
root@21b14f4a4c40:/tmp/cpdf-source# file ./cpdf
./cpdf: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=490fc74a859c8c6931b0ab1aa0d207abaa092e2e, not stripped
```

And, just as it may help somehow, the output of `file <faster_cpdf>`:
```
root@21b14f4a4c40:/tmp/cpdf-source# file ../cpdf
../cpdf: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=cec462743cfda5b86c8bc7b2e2f9f9ffacc88b89, not stripped
```

**Comment by johnwhitington** [comment]
Fixed in forthcoming v2.3. Bookmark names for @B are stripped as before, unless -utf8 is supplied, in which case problematic characters and characters < 32 only are stripped. If -raw is supplied, the text is not processed at all (not recommended).
